### PR TITLE
Update `spack.location` to be machine-readable

### LINKED ISSUE #51: Machine readable spack.location file in releases
A machine readable version of `spack.location` in the release artefacts would be useful for checking exe paths

https://github.com/ACCESS-NRI/access-om2-configs/issues/77#issuecomment-2060466317

and might also be useful for a model build DB

https://github.com/ACCESS-NRI/build-cd/issues/33

### PR DESCRIPTION
**Pull request by CodeGat**
In this PR:
* deploy-2-start.yml: Updated `spack.location` to be machine-readable 

Closes #51

### CODE REVIEW

**Comment by aidanheerdegen** [comment]
Can you paste an example of what it will look like in this PR? 

I'm wondering how readable it is, if we need a `spack.location` and `spack.location.json`

**Comment by CodeGat** [comment]
Was just going to do that. Seems that it ignores the `--path` bit if there's `--json`. Will think up a different way to do it...

**Comment by aidanheerdegen** [comment]
> Seems that it ignores the `--path` bit if there's `--json`. Will think up a different way to do it...

It's ok, if it dumps a whole heap of info into the `json` version we can live it it, and provide a clean text version as well.

**Comment by CodeGat** [comment]
If only. The dumped `json` doesn't have path information :D 
Here is one option `spack find --format '{name}@{version} {prefix}'`:
```
openssh@9.5p1 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/openssh-9.5p1-qkquurb34tyf46cdr2nhmoc6oth75z6m
openssl@3.1.3 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/openssl-3.1.3-ndunuvkldvcazhl4mfzsepm4tsnizoz4
patchelf@0.17.2 /opt/release/linux-rocky8-x86_64/gcc-8.5.0/patchelf-0.17.2-tr5ogxrot3gtfebg2un5dg7w2plh7kar
perl@5.26.3 /usr
pigz@2.7 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/pigz-2.7-twew7r5ag3dmmxkuozcjojxk4kuimdfa
pkgconf@1.9.5 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/pkgconf-1.9.5-ip67lrr46s2v6tvtljgitlf575xistht
pmix@4.2.2 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/pmix-4.2.2-dsuhyqk3l7lc44tzplnroid5wnfuizsm
snappy@1.1.10 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/snappy-1.1.10-jfjts53dpbxkx54nlmtjjqaaizrt3xb6
tar@1.34 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/tar-1.34-qdhx2ratkxgo66gw6m3cxohygy7wvn37
util-macros@1.19.3 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/util-macros-1.19.3-b6456jvl7wbyebhcihnkkpfdbwowejb5
xz@5.4.1 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/xz-5.4.1-b3nbkp7pvvondi5h6xvvoyiba44t4wl4
zlib-ng@2.1.4 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/zlib-ng-2.1.4-kg2thruurxgpfranmaxvrytyjlenhivb
zstd@1.5.5 /opt/release/linux-rocky8-x86_64/intel-2021.6.0/zstd-1.5.5-ntyz4di6t4unkyha2q6a7trvshtnpyxj
```
I'll see if I can pipe that into a `json` format

**Comment by CodeGat** [comment]
Oh but wait, theres a better solution: `spack find --format '{name}@{version} {prefix}' | jq -Rn '[inputs | split(" ") | {(.[0]):(.[1])}] | add'`, which gives:
```json
{
  "autoconf@2.69": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/autoconf-2.69-tma4iu2myuj6flenh44h2bcjmhi3ec4x",
  "automake@1.16.5": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/automake-1.16.5-mhwhddwrpsa6pnyvvp6lhlgm4vigrl3d",
  "bison@3.8.2": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/bison-3.8.2-cychjn2by2kilhp3siwwxojvlrqcueup",
  "bzip2@1.0.8": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/bzip2-1.0.8-qznl3hs3bgldp6hkwes2syduwz7kl4ol",
  "c-blosc@1.21.5": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/c-blosc-1.21.5-abehl6zew3lfiionlswhv3v74keteofn",
  "ca-certificates-mozilla@2023-05-30": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/ca-certificates-mozilla-2023-05-30-yr5vewcsaxxujzdczqfzazeiyy5ok22s",
  "cmake@3.24.2": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/cmake-3.24.2-aflo7vsxi74c25ulfctpapnn54wnmnhk",
  "curl@8.4.0": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/curl-8.4.0-cljy3dcvcaozkyw4uv7fdeqzwrh2qzwg",
  "datetime-fortran@1.7.0": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/datetime-fortran-1.7.0-4ynwokctw3a6jp762lpmtpcnfh7us6kk",
  "diffutils@3.9": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/diffutils-3.9-6sulf26ysg3mitdrnzqpn4le726rllnv",
  "findutils@4.9.0": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/findutils-4.9.0-k5jtb4lzzurfpksfymp363npw42y7mvs",
  "gettext@0.22.3": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/gettext-0.22.3-zrbrnowc4jaxjmd27bif6pi2fzny7jlz",
  "gmake@4.4.1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/gmake-4.4.1-r6gn6i6hnqiiju2ja6vnh74cw7y7ku2t",
  "hdf5@1.14.3": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/hdf5-1.14.3-6d5kw3irv4ufih5pg5q7hp2tx6zvadi3",
  "hwloc@2.9.1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/hwloc-2.9.1-eyartwswmwd3uloutodpi7dv2ftimumg",
  "intel-oneapi-compilers@2022.1.0": "/opt/release/linux-rocky8-x86_64/gcc-8.5.0/intel-oneapi-compilers-2022.1.0-rsloualixgaqwytkvyjeujo33na5hl74",
  "intel-oneapi-compilers-classic@2021.6.0": "/opt/release/linux-rocky8-x86_64/gcc-8.5.0/intel-oneapi-compilers-classic-2021.6.0-pztudm2akp43tlxnwxbzfop77a6drsro",
  "json-fortran@8.3.0": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/json-fortran-8.3.0-fzbed3rqp5yvh2mbwa2qdmy5juqw4err",
  "krb5@1.20.1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/krb5-1.20.1-rmzero3wmrkodeajbkdefxmkxdonbbb7",
  "libaccessom2@master": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libaccessom2-master-oyd7uzkdrkuhpji7tluldtvgn63mcbte",
  "libaec@1.0.6": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libaec-1.0.6-nyo6cccbjttq2thltiwqbw65odebnbez",
  "libedit@3.1-20210216": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libedit-3.1-20210216-qm5xdopluxdp7bji2vn5oljq3vsqnycd",
  "libevent@2.1.12": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libevent-2.1.12-rbdqr5wpdphlbi2g34vvv37qgeihkcen",
  "libiconv@1.17": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libiconv-1.17-3y6mhfzdi6tz2mxjl27eks5dkn435vgx",
  "libpciaccess@0.17": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libpciaccess-0.17-7nxupyhibb6w6fgsybuev7hezoc623rk",
  "libsigsegv@2.14": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libsigsegv-2.14-s2r72h6humup6hy5aqrme4yfhnppdail",
  "libtool@2.4.6": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libtool-2.4.6-4n4envtiooeppllgz7av3hbxgryqjfv4",
  "libxcrypt@4.4.35": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libxcrypt-4.4.35-2o2paughoi5pqad3g6sxedrgxshqnqkc",
  "libxml2@2.10.3": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/libxml2-2.10.3-micdp6afcr65o4wmtydwv4ay3htz6bak",
  "lz4@1.9.4": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/lz4-1.9.4-6h7im7753qzyx6s43dzd7mrroc5cm3qv",
  "m4@1.4.19": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/m4-1.4.19-ty7t3fe7fzjd6xbbeifvc6hg4apx2hcm",
  "mom5@master": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/mom5-master-zoloemhdjgywvcpyodimfikpyibe74yz",
  "ncurses@6.4": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/ncurses-6.4-rbg6b4chtj2ibyygqdvyhl5im74c4jre",
  "netcdf-c@4.9.2": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/netcdf-c-4.9.2-2v4k34wjvlnrqyzldby5rhun5jcvh47j",
  "netcdf-fortran@4.6.1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/netcdf-fortran-4.6.1-547f5njbyicc3ap3hzwkik7m6bjwbvqw",
  "nghttp2@1.57.0": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/nghttp2-1.57.0-exglwvj2veznp7urmwvwvreaueectufy",
  "numactl@2.0.14": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/numactl-2.0.14-eg2jpug6stndspge2vyqxp4lt5qcx62q",
  "oasis3-mct@master": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/oasis3-mct-master-32b5mxdjrk3fr6ivla6a2g3vzfr2ll4l",
  "openmpi@4.0.2": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/openmpi-4.0.2-c6w7gubuqlurxqxdmbdyekswwxbcz7ll",
  "openssh@9.5p1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/openssh-9.5p1-qkquurb34tyf46cdr2nhmoc6oth75z6m",
  "openssl@3.1.3": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/openssl-3.1.3-ndunuvkldvcazhl4mfzsepm4tsnizoz4",
  "patchelf@0.17.2": "/opt/release/linux-rocky8-x86_64/gcc-8.5.0/patchelf-0.17.2-tr5ogxrot3gtfebg2un5dg7w2plh7kar",
  "perl@5.26.3": "/usr",
  "pigz@2.7": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/pigz-2.7-twew7r5ag3dmmxkuozcjojxk4kuimdfa",
  "pkgconf@1.9.5": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/pkgconf-1.9.5-ip67lrr46s2v6tvtljgitlf575xistht",
  "pmix@4.2.2": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/pmix-4.2.2-dsuhyqk3l7lc44tzplnroid5wnfuizsm",
  "snappy@1.1.10": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/snappy-1.1.10-jfjts53dpbxkx54nlmtjjqaaizrt3xb6",
  "tar@1.34": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/tar-1.34-qdhx2ratkxgo66gw6m3cxohygy7wvn37",
  "util-macros@1.19.3": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/util-macros-1.19.3-b6456jvl7wbyebhcihnkkpfdbwowejb5",
  "xz@5.4.1": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/xz-5.4.1-b3nbkp7pvvondi5h6xvvoyiba44t4wl4",
  "zlib-ng@2.1.4": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/zlib-ng-2.1.4-kg2thruurxgpfranmaxvrytyjlenhivb",
  "zstd@1.5.5": "/opt/release/linux-rocky8-x86_64/intel-2021.6.0/zstd-1.5.5-ntyz4di6t4unkyha2q6a7trvshtnpyxj"
}
```

File: .github/workflows/deploy-2-start.yml [hunk 1]
```diff
@@ -73,7 +73,7 @@ jobs:
           spack env activate ${{ inputs.env-name }}
           spack --debug install --fresh || exit $?
           spack module tcl refresh --delete-tree -y
-          spack find --paths > ${{ vars.SPACK_LOCATION }}/var/spack/environments/${{ inputs.env-name }}/spack.location
+          spack find --format '{name}@{version} {prefix}' | jq -Rn '[inputs | split(" ") | {(.[0]):(.[1])}] | add' > ${{ vars.SPACK_LOCATION }}/var/spack/environments/${{ inputs.env-name }}/spack.location
```
Comment (aidanheerdegen) [inline .github/workflows/deploy-2-start.yml:76]: It is great to have a machine-readable version. Can (should) we add a `.json` suffix? Would that make it open in a browser if clicked, or does it always force a download?

What do you think about having both?

**Review by aidanheerdegen — DISMISSED** [review]

File: .github/workflows/deploy-2-start.yml (hunk 1 above)
Comment (CodeGat) [reply .github/workflows/deploy-2-start.yml:76]: Good point. Lets do both!

Comment (aidanheerdegen) [inline .github/workflows/deploy-2-start.yml]: So what happens if there is normal text and `json` formatted text appended? Is that parseable?

**Review by aidanheerdegen — APPROVED** [review]
Looks good, just another question.

**Review by aidanheerdegen — CHANGES_REQUESTED** [review]
Actually, I think that was a mistake to not use `spack.location.json` because it is overwritten as currently written

File: .github/workflows/deploy-2-start.yml (hunk 1 above)
Comment (CodeGat) [reply .github/workflows/deploy-2-start.yml]: One is `spack.location` and one is `spack.location.json` - they won't be appended together:
https://github.com/ACCESS-NRI/build-cd/pull/52/files#diff-8984338bca7c293efa309256806ef1822893d031812c36f28b661d9f4b142383R76-R77

File: .github/workflows/deploy-2-start.yml [hunk 2]
```diff
@@ -74,6 +74,7 @@ jobs:
           spack --debug install --fresh || exit $?
           spack module tcl refresh --delete-tree -y
           spack find --paths > ${{ vars.SPACK_LOCATION }}/var/spack/environments/${{ inputs.env-name }}/spack.location
+          spack find --format '{name}@{version} {prefix}' | jq --raw-input --null-input '[inputs | split(" ") | {(.[0]):(.[1])}] | add' > ${{ vars.SPACK_LOCATION }}/var/spack/environments/${{ inputs.env-name }}/spack.location.json
```
Comment (aidanheerdegen) [inline .github/workflows/deploy-2-start.yml:77]: WUNDERBAR!

**Review by aidanheerdegen — APPROVED** [review]
My diff was as stale as week old bread. 

Sorry for the mixed messages.

**Pull request merged**